src/metabase/api/search.clj

```diff
@@ -246,60 +246,70 @@
 
 (defmulti ^:private search-query-for-model
   {:arglists '([model search-context])}
-  (fn [model _] (class model)))
+  (fn [model _] model))
 
-(s/defmethod search-query-for-model (class Card)
-  [_ search-ctx :- SearchContext]
-  (-> (base-query-for-model Card search-ctx)
+(s/defn ^:private shared-card-impl [dataset? :- s/Bool search-ctx :- SearchContext]
+  (-> (base-query-for-model "card" search-ctx)
+      (update :where (fn [where] [:and [:= :card.dataset dataset?] where]))
       (h/left-join [CardFavorite :fave]
                    [:and
                     [:= :card.id :fave.card_id]
                     [:= :fave.owner_id api/*current-user-id*]])
       (add-collection-join-and-where-clauses :card.collection_id search-ctx)
       (add-card-db-id-clause (:table-db-id search-ctx))))
 
-(s/defmethod search-query-for-model (class Collection)
-  [_ search-ctx :- SearchContext]
-  (-> (base-query-for-model Collection search-ctx)
+(s/defmethod search-query-for-model "card"
+  [_model search-ctx :- SearchContext]
+  (shared-card-impl false search-ctx))
+
+(s/defmethod search-query-for-model "dataset"
+  [_model search-ctx :- SearchContext]
+  (-> (shared-card-impl true search-ctx)
+      (update :select (fn [columns]
```

Comment:
it looks like you could do something with `merge-select` on hsql instead of this one? OK if you don't, just noting

Reply:
I don't think that i can because i'm removing the first column and replacing it. The card query returns `select 'card' as model` and i need to update that here to `select 'dataset' as model`. Threading that all the way down was a pain so I just take the columns, toss away the first, and add a new model clause.